File: brand-update-product-create-race/python/reconcile_brand_product_race.py
"""Reconcile a BigCommerce brand update fired immediately before a product create
that came back as an empty reply from server.

BigCommerce enforces a per-store request quota (150 to 450 requests per 30 second
OAuth window depending on plan) and a concurrency cap, normally surfaced as a 429
with X-Rate-Limit-Requests-Left and X-Rate-Limit-Time-Reset-Ms headers. When a
brand PUT to /v3/catalog/brands/{id} is fired immediately before a product POST to
/v3/catalog/products, the store's connection sometimes closes before the response
finishes, which HTTP clients surface as a generic empty reply instead of a
structured error. The underlying mutation may have actually succeeded server side
even though the client received nothing parsable. This is a confirmed, reproduced
issue in BigCommerce's own bigcommerce-api-php SDK repo (issue #138).

This job takes logged (brand_id, intended_fields, product_payload) pairs, confirms
whether the brand update actually applied, checks whether the product already
exists from the failed attempt, and only retries the create when it is safe:
brand confirmed, product confirmed absent, and rate limit budget or backoff
allows another call. Anything else (a stale brand update, or a same-named product
whose fields do not match) is flagged for manual review, never auto-repaired.

Guide: https://www.allanninal.dev/bigcommerce/brand-update-product-create-race/
"""
import os
import time
import logging

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("reconcile_brand_product_race")

STORE_HASH = os.environ["BIGCOMMERCE_STORE_HASH"]
ACCESS_TOKEN = os.environ["BIGCOMMERCE_ACCESS_TOKEN"]
API_BASE = f"https://api.bigcommerce.com/stores/{STORE_HASH}/v3"
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
MAX_ATTEMPTS = int(os.environ.get("MAX_ATTEMPTS", "5"))

HEADERS = {
    "X-Auth-Token": ACCESS_TOKEN,
    "Content-Type": "application/json",
    "Accept": "application/json",
}


def bc_get(path, params=None):
    r = requests.get(f"{API_BASE}{path}", headers=HEADERS, params=params or {}, timeout=30)
    r.raise_for_status()
    rate_limit_left = int(r.headers.get("X-Rate-Limit-Requests-Left", "1"))
    return r.json(), rate_limit_left


def bc_post(path, body):
    r = requests.post(f"{API_BASE}{path}", headers=HEADERS, json=body, timeout=30)
    r.raise_for_status()
    return r.json()


def decide_action(
    brand_confirmed: bool,
    product_exists: bool,
    rate_limit_left: int,
    attempt: int,
    max_attempts: int = 5,
) -> str:
    """Pure decision logic, no I/O. Returns one of:

    'noop_success'        - brand_confirmed and product_exists: pair actually
                             succeeded despite empty reply.
    'retry_create'        - brand_confirmed, not product_exists, rate_limit_left
                             > 0, attempt < max_attempts: safe to retry create.
    'wait_and_retry'      - rate_limit_left <= 0 and attempt < max_attempts:
                             back off before retrying.
    'flag_manual_review'  - not brand_confirmed (brand update itself never
                             applied): don't create against a stale brand.
    'give_up'             - attempt >= max_attempts and not product_exists:
                             surface for manual review.
    """
    if brand_confirmed and product_exists:
        return "noop_success"
    if not brand_confirmed:
        return "flag_manual_review"
    if attempt >= max_attempts and not product_exists:
        return "give_up"
    if rate_limit_left <= 0:
        return "wait_and_retry"
    return "retry_create"


def confirm_brand_update(brand_id, intended_fields):
    data, rate_limit_left = bc_get(f"/catalog/brands/{brand_id}")
    brand = data.get("data", {})
    matches = all(brand.get(field) == value for field, value in intended_fields.items())
    return matches, rate_limit_left


def find_existing_product(name, brand_id):
    data, rate_limit_left = bc_get("/catalog/products", {"name": name, "brand_id": brand_id})
    products = data.get("data", [])
    return (products[0] if products else None), rate_limit_left


def create_product(product_payload):
    return bc_post("/catalog/products", product_payload)


def backoff_seconds(attempt):
    return min(2 ** attempt, 8)


def reconcile_pair(brand_id, intended_fields, product_payload):
    attempt = 0
    while True:
        brand_confirmed, rate_limit_left = confirm_brand_update(brand_id, intended_fields)
        existing, rate_limit_left = find_existing_product(
            product_payload.get("name"), brand_id
        )
        product_exists = existing is not None

        decision = decide_action(brand_confirmed, product_exists, rate_limit_left, attempt, MAX_ATTEMPTS)

        log.info(
            "brand_id=%s attempt=%s brand_confirmed=%s product_exists=%s "
            "rate_limit_left=%s decision=%s",
            brand_id, attempt, brand_confirmed, product_exists, rate_limit_left, decision,
        )

        if decision == "noop_success":
            return "noop_success"
        if decision == "flag_manual_review":
            log.warning("Brand %s update not confirmed. Flagging pair for manual review.", brand_id)
            return "flag_manual_review"
        if decision == "give_up":
            log.warning("Brand %s exhausted %s attempts. Flagging for manual review.", brand_id, MAX_ATTEMPTS)
            return "give_up"
        if decision == "wait_and_retry":
            wait_for = backoff_seconds(attempt)
            log.info("Rate limit exhausted, waiting %ss before retry.", wait_for)
            if not DRY_RUN:
                time.sleep(wait_for)
            attempt += 1
            continue

        # decision == "retry_create"
        if DRY_RUN:
            log.info("Dry run: would create product %s under brand %s.", product_payload.get("name"), brand_id)
            return "retry_create"

        # Re-check existence immediately before writing, state can change between checks.
        existing_recheck, _ = find_existing_product(product_payload.get("name"), brand_id)
        if existing_recheck is not None:
            log.info("Product appeared before retry. Treating as noop_success.")
            return "noop_success"

        create_product(product_payload)
        log.info("Created product %s under brand %s.", product_payload.get("name"), brand_id)
        return "created"


def run(pairs):
    """pairs: iterable of (brand_id, intended_fields, product_payload)."""
    results = []
    for brand_id, intended_fields, product_payload in pairs:
        results.append(reconcile_pair(brand_id, intended_fields, product_payload))
    log.info("Done. %d pair(s) processed.", len(results))
    return results


if __name__ == "__main__":
    run([])

File: brand-update-product-create-race/node/reconcile-brand-product-race.js
/**
 * Reconcile a BigCommerce brand update fired immediately before a product create
 * that came back as an empty reply from server.
 *
 * BigCommerce enforces a per-store request quota (150 to 450 requests per 30
 * second OAuth window depending on plan) and a concurrency cap, normally
 * surfaced as a 429 with X-Rate-Limit-Requests-Left and
 * X-Rate-Limit-Time-Reset-Ms headers. When a brand PUT to
 * /v3/catalog/brands/{id} is fired immediately before a product POST to
 * /v3/catalog/products, the store's connection sometimes closes before the
 * response finishes, which HTTP clients surface as a generic empty reply
 * instead of a structured error. The underlying mutation may have actually
 * succeeded server side even though the client received nothing parsable.
 * This is a confirmed, reproduced issue in BigCommerce's own bigcommerce-api-php
 * SDK repo (issue #138).
 *
 * This job takes logged {brandId, intendedFields, productPayload} pairs,
 * confirms whether the brand update actually applied, checks whether the
 * product already exists from the failed attempt, and only retries the create
 * when it is safe: brand confirmed, product confirmed absent, and rate limit
 * budget or backoff allows another call. Anything else is flagged for manual
 * review, never auto-repaired.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/brand-update-product-create-race/
 */
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "example_hash";
const ACCESS_TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "bc_dummy";
const API_BASE = `https://api.bigcommerce.com/stores/${STORE_HASH}/v3`;
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const MAX_ATTEMPTS = Number(process.env.MAX_ATTEMPTS || 5);

const HEADERS = {
  "X-Auth-Token": ACCESS_TOKEN,
  "Content-Type": "application/json",
  Accept: "application/json",
};

/**
 * Pure decision logic, no I/O. Returns one of:
 *
 * "noop_success"        - brandConfirmed and productExists: pair actually
 *                          succeeded despite empty reply.
 * "retry_create"        - brandConfirmed, not productExists, rateLimitLeft
 *                          > 0, attempt < maxAttempts: safe to retry create.
 * "wait_and_retry"      - rateLimitLeft <= 0 and attempt < maxAttempts:
 *                          back off before retrying.
 * "flag_manual_review"  - not brandConfirmed (brand update itself never
 *                          applied): don't create against a stale brand.
 * "give_up"             - attempt >= maxAttempts and not productExists:
 *                          surface for manual review.
 */
export function decideAction(brandConfirmed, productExists, rateLimitLeft, attempt, maxAttempts = 5) {
  if (brandConfirmed && productExists) return "noop_success";
  if (!brandConfirmed) return "flag_manual_review";
  if (attempt >= maxAttempts && !productExists) return "give_up";
  if (rateLimitLeft <= 0) return "wait_and_retry";
  return "retry_create";
}

async function bcGet(path, params = {}) {
  const url = new URL(`${API_BASE}${path}`);
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null) url.searchParams.set(key, value);
  }
  const res = await fetch(url, { headers: HEADERS });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  const rateLimitLeft = Number(res.headers.get("X-Rate-Limit-Requests-Left") || "1");
  const body = await res.json();
  return { body, rateLimitLeft };
}

async function bcPost(path, body) {
  const res = await fetch(`${API_BASE}${path}`, {
    method: "POST",
    headers: HEADERS,
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  return res.json();
}

async function confirmBrandUpdate(brandId, intendedFields) {
  const { body, rateLimitLeft } = await bcGet(`/catalog/brands/${brandId}`);
  const brand = body.data || {};
  const matches = Object.entries(intendedFields).every(([field, value]) => brand[field] === value);
  return { matches, rateLimitLeft };
}

async function findExistingProduct(name, brandId) {
  const { body, rateLimitLeft } = await bcGet("/catalog/products", { name, brand_id: brandId });
  const products = body.data || [];
  return { product: products[0] || null, rateLimitLeft };
}

async function createProduct(productPayload) {
  return bcPost("/catalog/products", productPayload);
}

function backoffSeconds(attempt) {
  return Math.min(2 ** attempt, 8);
}

function sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

async function reconcilePair(brandId, intendedFields, productPayload) {
  let attempt = 0;
  for (;;) {
    const { matches: brandConfirmed } = await confirmBrandUpdate(brandId, intendedFields);
    const { product: existing, rateLimitLeft } = await findExistingProduct(productPayload.name, brandId);
    const productExists = existing !== null;

    const decision = decideAction(brandConfirmed, productExists, rateLimitLeft, attempt, MAX_ATTEMPTS);

    console.log(
      `brand_id=${brandId} attempt=${attempt} brand_confirmed=${brandConfirmed} ` +
      `product_exists=${productExists} rate_limit_left=${rateLimitLeft} decision=${decision}`
    );

    if (decision === "noop_success") return "noop_success";
    if (decision === "flag_manual_review") {
      console.warn(`Brand ${brandId} update not confirmed. Flagging pair for manual review.`);
      return "flag_manual_review";
    }
    if (decision === "give_up") {
      console.warn(`Brand ${brandId} exhausted ${MAX_ATTEMPTS} attempts. Flagging for manual review.`);
      return "give_up";
    }
    if (decision === "wait_and_retry") {
      const waitFor = backoffSeconds(attempt);
      console.log(`Rate limit exhausted, waiting ${waitFor}s before retry.`);
      if (!DRY_RUN) await sleep(waitFor);
      attempt += 1;
      continue;
    }

    // decision === "retry_create"
    if (DRY_RUN) {
      console.log(`Dry run: would create product ${productPayload.name} under brand ${brandId}.`);
      return "retry_create";
    }

    const { product: recheck } = await findExistingProduct(productPayload.name, brandId);
    if (recheck !== null) {
      console.log("Product appeared before retry. Treating as noop_success.");
      return "noop_success";
    }

    await createProduct(productPayload);
    console.log(`Created product ${productPayload.name} under brand ${brandId}.`);
    return "created";
  }
}

export async function run(pairs = []) {
  const results = [];
  for (const { brandId, intendedFields, productPayload } of pairs) {
    results.push(await reconcilePair(brandId, intendedFields, productPayload));
  }
  console.log(`Done. ${results.length} pair(s) processed.`);
  return results;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run([]).catch((err) => { console.error(err); process.exit(1); });
}
